### ISSUE TITLE
Problem of receiving two or more duplicate messages when sending a message through chat_postMessage()

### ISSUE THREAD

**Issue by sh4n3e** [slackapi/bolt-python#764]
Problem of receiving two or more duplicate messages when sending a message through chat_postMessage()

I'm using slack_bolt and when I send a message through chat_postMessage(), I get a problem with receiving a message with two or more duplicate messages.
The message is sent as a combination of blocks, and the size of the block exceeds 4000 bytes. I tried debugging, but it doesn't seem to retry.
 An example of a block is shown below.
```
ex1) app.client.chat_postMessage(blocks=, channel=)
ex2) blocks = [{'elements': [{'text': '*aaaaaaaaaa*', 'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'alt_text': 'cute cat',
                           'image_url': 'https://cdn-icons-png.flaticon.com/512/752/752664.png',
                           'type': 'image'},
                          {'text': f'{"a"*37}',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'text': f'>```{"a"*3920}```',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'alt_text': 'cute cat',
                           'image_url': 'https://cdn-icons-png.flaticon.com/512/752/752665.png',
                           'type': 'image'},
                          {'text': f'{"a"*39}',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'text': f'>```{"a"*100}```', 'type': 'mrkdwn'}],
             'type': 'context'}]

```

**Comment by WilliamBergamin** [comment]
Hi @sh4n3e thanks for witting in!

In order to try and reproduce this on my end would it be possible for you to share a snippet of your source code, and what the duplicate messages looks like.

But here are 2 things to look out for:
- Depending on the handler you are implementing, ensure that the `ack()` function is called before `client.chat_postMessage` similar [to this example](https://slack.dev/bolt-python/concepts#shortcuts) where `ack()` is called before `client.views_open`.
- Ensure that your handler is not hanging indefinitely, a retry request could be sent to your app if no response is received

**Comment by sh4n3e** [comment]
Hi @WilliamBergamin 

An example of a block is shown below.
I'm using it differently from the example you gave me.

```
@app.message("test")
        def testfunc(message):
            if message['text'] == 'test':
                app.client.chat_postMessage(blocks=testtemp(), channel=message['channel'])
```
```
def testtemp() -> list:
    return [{'elements': [{'text': '*aaaaaaaaaa*', 'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'alt_text': 'cute cat',
                           'image_url': 'https://cdn-icons-png.flaticon.com/512/752/752664.png',
                           'type': 'image'},
                          {'text': f'{"a"*37}',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'text': f'>```{"a"*3920}```',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'alt_text': 'cute cat',
                           'image_url': 'https://cdn-icons-png.flaticon.com/512/752/752665.png',
                           'type': 'image'},
                          {'text': f'{"a"*39}',
                           'type': 'mrkdwn'}],
             'type': 'context'},
            {'elements': [{'text': f'>```{"a"*100}```', 'type': 'mrkdwn'}],
             'type': 'context'}]
```

**Comment by seratch** [comment]
@sh4n3e The only possible direct cause that we can think of is that your bolt-python app receives the message events several times, so that the same chat.postMessage API call is executed twice or more. The common reason of this is that your bolt-python app does not respond to message event payload data request from Slack within 3 seconds for some reason and then the Slack API server retries the same request up to 3 times.

In the case of `app.event` / `app.message` listeners, you don't need to explicitly call `ack()` method within 3 seconds in the listeners. bolt-python framework does `ack()` for you under the hood. However, if any of the following situations arises, acknowledging within 3 seconds may fail:
* Your app have `process_before_response=True` in `App` constructor for FaaS compatibility and your listener takes more than 3 seconds to complete
* Your app received message event requests but failed to send an HTTP response to it for some reason
* If you use Socket Mode, your WebSocket connection might be unstable for a reason such as network glitch

To figure out what's happening on your app, I would suggest enabling debug-level logging for your bolt-python app. Please refer to https://slack.dev/bolt-python/concepts#logging to learn how to configure loggers. Also, if you have access logs on the load balancer or frontend web server (e.g., nginx, Apache HTTP server), checking whether message event requests are properly responded with 200 OK and how long the server takes to respond should be helpful for you.

I hope that you will find the solution for it soon!

**Comment by ls-urs-keller** [comment]
I have the same issue with a and can reproduce it this way:

```
curl -v -H "Content-type: application/json; charset=utf-8" \
  --data "@trial.json"  \
  -H "Authorization: Bearer ${SLACK_ACCESS_TOKEN}" \
  -X POST https://slack.com/api/chat.postMessage
```
[trial.json.zip](https://github.com/slackapi/bolt-python/files/11200112/trial.json.zip)

Using this curl you will see that the messages are duplicated in the channel. 

If you remove 1 character from the end of the message (remove the `5`) there are no duplicates. If you create even bigger messages you get 3 or 4 times the same message.

This must be related to the slack API and how it handles sizes.
@seratch I can post this in a different place this doesn't seem to be linked to the client library, but is a bug in the slack API.
We observe this since the 30th March 2023.

**Comment by g-bakas** [comment]
We have been experiencing the same behaviour when posting big slack messages. The message is posted twice in the slack channel, even though smaller messages are posted once. There are no code changes (our code had been working properly since a couple of months ago), but the messaging behaviour changed over the past weeks. We are using the _reqeusts_ python library as follows: 


`def send(self, webhook):

        payload = {
            "blocks": self._components
        }

        requests.post(webhook, json.dumps(payload))`

**Comment by kacper1112** [comment]
I reached out to Slack's support about this and got a response:

>This is a known bug with messages which are over 4039 characters of text combined. The expected behaviour is that messages over 4000 characters should get split up and posted across multiple messages, up to a maximum of 40,000 characters (which should be split into up to 10 messages.)(https://api.slack.com/changelog/2018-04-truncating-really-long-messages / https://api.slack.com/changelog/2018-04-truncating-really-long-messages). It seems our logic on splitting up these messages is causing duplicates of the first message to be posted instead.

>I'm afraid we don't have a timeline of how long it may take to resolve this issue, but I shall be sure to update you here as soon as there's any word on it. In the meantime I would recommend splitting up any messages longer than 4000 characters on your end before sending.

Posting this here as I see more users are looking for some information on the problem.

**Comment by alfa-jpn** [comment]
@ls-urs-keller @colinmcparland @g-bakas @kacper1112 

Oddly enough, I discovered that if both "blocks" and "text" were set, the messages were not duplicated.
I use it as a workaround.

eg)
```json
{
  channel: "xxx",
  text: "no-use",
  blocks: [~~~]
}
```

**Comment by dmytrostrembetskyi** [comment]
thanks, @alfa-jpn 
specifying the `text` resolved the issue 
```
    public async Task<PostMessageResponse> PostMessageAsync(string userId, IBlock[] blocks)
    {
        return await this.slackTaskClient.PostMessageAsync(
                        userId,
                        "no-use",
                        null,
                        null,
                        false,
                        blocks);
    }
```

**Comment by eddyg** [comment]
It is worth noting that the `text` attribute is **not** ignored; it is used when displaying notifications (and potentially for by screen readers for vision-impaired Slack users and other alternative Slack interfaces).

Quoting the [docs](https://api.slack.com/methods/chat.postMessage):

> The formatted text of the message to be published. If blocks are included, this will become the fallback text used in notifications.

When calling `chat.postMessage` with blocks, the top-level `text` field should include a text-only representation of the blocks.

**Comment by ls-urs-keller** [comment]
> @ls-urs-keller @colinmcparland @g-bakas @kacper1112
> 
> Oddly enough, I discovered that if both "blocks" and "text" were set, the messages were not duplicated. I use it as a workaround.
@alfa-jpn thank you, perfect, works for my purposes just tested it.

**Comment by phil000** [comment]
Same problem happening for us.

We post messages to ``api/chat.postMessage`` and the same message appears in the Slack channel multiple times.

**Comment by mahsirat-atiye** [comment]
We have same problem
`import logging
import os
from time import sleep

from slack_bolt.async_app import AsyncApp

logging.basicConfig(level=logging.DEBUG)
app = AsyncApp(
    token="xoxb-....",
    signing_secret="..."
)


@app.event("app_mention")
async def event_test(body, say, logger):
    logger.info(body)
    await say("What's up?")


@app.event("message")
async def mention_handler(body, say):
    # Get the message sent by the user
    await respond_to_new_message(body["event"], say, app.client)


async def respond_to_new_message(message, say, client):
    # Get the text from the message the user sent
    text = message.get("text")
    # Get the channel id from the message the user sent
    channel_id = message["channel"]
    # Respond to the user
    blocks = [
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": f"You said {text}"
            }
        }
    ]
    # Post the response message to the channel
    await client.chat_postMessage(channel=channel_id, blocks=blocks)

    # Post a message with the answer
    ans = await find_answer_using_gbt(text)
    answer_blocks = [
        {
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": ans
            }
        }
    ]
    await client.chat_postMessage(channel=channel_id, blocks=answer_blocks)


async def find_answer_using_gbt(text):
    # ...
    sleep(10)
    return "gbt answer to {}".format(text)


from fastapi import FastAPI, Request

api = FastAPI()

# to make it able to response to slack challenge requests
# 1. we create a challenge handler
# 2. we create a challenge endpoint

from slack_bolt.adapter.fastapi.async_handler import AsyncSlackRequestHandler

app_handler = AsyncSlackRequestHandler(app)


@api.post("/slack/events")
async def endpoint(req: Request):
    return await app_handler.handle(req)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(api, host="0.0.0.0", port=80)
`

**Comment by WilliamBergamin** [comment]
@mahsirat-atiye thanks for the detailed example 💯 

could you try adding `ack()` in `mention_handler` it could look something like this
```python
@app.event("message")
async def mention_handler(body, say, ack):
    await ack()
    # Get the message sent by the user
    await respond_to_new_message(body["event"], say, app.client)
```

**Comment by seratch** [comment]
@mahsirat-atiye This may not be the cause of the issue you mentioned here but generally speaking, `time.sleep` in an asyncio app can be a showstopper. You must avoid it in any case. Instead, you can use `await asyncio.sleep(duration)`. Refer to https://github.com/slackapi/bolt-python/issues/791#issuecomment-1355825843 for details.

**Comment by tadeodonegana** [comment]
I am experiencing the same issue. Any news on this @seratch ? 

Thanks in advance.

**Comment by tofsjonas** [comment]
Same 😞

**Comment by pawurb** [comment]
Same for me as well 😞

**Comment by etandon-coatue** [comment]
Seeing the same issue with the direct slack API call, how do you send an ack() to the slack API?

**Comment by ashrekhi** [comment]
facing the same issue, anyone able to fix this?

**Comment by la0bing** [comment]
also hitting this issue occasionally here

**Comment by jaoassis** [comment]
hey guys! i'v seen that few of u had the same issue that i had, related to use slackbolt and using mentions and interactions where multiple messages are sent in a channel or person. 
reading the slack's documentation, they mention that they send retries in few seconds, which can cause the multiple messages in case of the application executes each time that receive the POST from slack: https://api.slack.com/apis/events-api#retries

**what i did to solve it:** when they send the retries, they send some headers, like x-slack-retry-reason and x-slack-retry-num, what is not sent in the first POST.
i basically created a condition which checks if the header x-slack-retry-reason not exist in the headers to proceed and execute the code:
in my case, im using lambda, but you can use the same logic in other cases, for example:

`if "headers" in event and "X-Slack-Retry-Reason" not in event["headers"]:
         SlackRequestHandler(slack_app).handle(event, context)`
        
i hope it helps! 😄
